pallets/gear-program/src/migrations.rs

```diff
@@ -0,0 +1,307 @@
+                |program_id, program: v2::Program<BlockNumberFor<T>>| {
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                    Some(match program {
+                        v2::Program::Active(p) => {
+                            for (page, data) in v2::MemoryPageStorage::<T>::drain_prefix(program_id)
+                            {
+                                weight =
+                                    weight.saturating_add(T::DbWeight::get().reads_writes(0, 2));
+
+                                MemoryPages::<T>::insert((program_id, MEMORY_INFIX, page), data);
+                            }
+
+                            Program::Active(common::ActiveProgram {
+                                allocations: p.allocations,
+                                pages_with_data: p.pages_with_data,
+                                gas_reservation_map: p.gas_reservation_map,
+                                code_hash: p.code_hash,
+                                code_exports: p.code_exports,
+                                static_pages: p.static_pages,
+                                state: p.state,
+                                expiration_block: p.expiration_block,
+                                memory_infix: MEMORY_INFIX,
+                            })
+                        }
+                        v2::Program::Exited(id) => Program::Exited(id),
+                        v2::Program::Terminated(id) => Program::Terminated(id),
+                    })
+                },
+            );
+
+            if v2::SessionMemoryPages::<T>::iter().next().is_some() {
+                log::error!("v2::SessionMemoryPages is not empty");
+            }
+
+            if ResumeSessions::<T>::iter().next().is_some() {
+                log::error!("ResumeSessions is not empty");
+            }
+
+            if PausedProgramStorage::<T>::iter().next().is_some() {
+                log::error!("PausedProgramStorage is not empty");
+            }
+
+            weight = weight.saturating_add(T::DbWeight::get().writes(1));
+            current.put::<Pallet<T>>();
+
+            log::info!("Successfully migrated storage");
+        } else {
+            log::info!("❌ Migration did not execute. This probably should be removed");
+        }
+
+        weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(state: Vec<u8>) -> Result<(), &'static str> {
+        // Check that everything decoded fine.
+        let count = ProgramStorage::<T>::iter_keys().fold(0u64, |i, k| {
+            let Ok(program) = ProgramStorage::<T>::try_get(k) else {
+                unreachable!("Cannot decode v3 Program");
+            };
+
+            if let Program::Active(p) = program {
+                assert_eq!(p.memory_infix, MEMORY_INFIX);
+
+                for page in p.pages_with_data.iter() {
+                    assert!(MemoryPages::<T>::contains_key((k, p.memory_infix, page)));
```

Comment:
off top, but please open an issue re review infix set on resume sessions for future changes